File: packages/lcyt/src/sender.js
import http from 'http';
import { URL } from 'url';
import { ConfigError, NetworkError, ValidationError } from './errors.js';
import logger from './logger.js';

// Default YouTube Live caption ingestion URL base
const DEFAULT_YOUTUBE_URL = 'http://upload.youtube.com/closedcaption';

/**
 * Client for sending live captions to YouTube Live streams using Google's closed caption API.
 * @class
 * @example
 * const sender = new YoutubeLiveCaptionSender({
 *   streamKey: 'YOUR_STREAM_KEY'
 * });
 * sender.start();
 * await sender.send('Hello, world!');
 * sender.end();
 */
class YoutubeLiveCaptionSender {
  /**
   * Create a new YoutubeLiveCaptionSender instance.
   * @param {Object} [options={}] - Configuration options
   * @param {string} [options.streamKey=null] - YouTube stream key (cid value)
   * @param {string} [options.baseUrl] - Base ingestion URL (defaults to YouTube's upload.youtube.com endpoint)
   * @param {string} [options.ingestionUrl=null] - Full pre-built ingestion URL (overrides streamKey and baseUrl)
   * @param {string} [options.region='reg1'] - Region identifier for captions
   * @param {string} [options.cue='cue1'] - Cue identifier for captions
   * @param {boolean} [options.useRegion=false] - Include region/cue in caption body
   * @param {number} [options.sequence=0] - Starting sequence number
   * @param {boolean} [options.verbose=false] - Enable verbose logging
   */
  constructor(options = {}) {
    this.streamKey = options.streamKey || null;
    this.baseUrl = options.baseUrl || DEFAULT_YOUTUBE_URL;
    this.region = options.region || 'reg1';
    this.cue = options.cue || 'cue1';
    this.useRegion = options.useRegion || false;
    this.sequence = options.sequence || 0;
    this.isStarted = false;
    this.verbose = options.verbose || false;
    this._queue = []; // Internal queue for construct/sendBatch pattern

    // Build ingestion URL: use provided ingestionUrl, or build from streamKey + baseUrl
    if (options.ingestionUrl) {
      this.ingestionUrl = options.ingestionUrl;
    } else if (this.streamKey) {
      this.ingestionUrl = `${this.baseUrl}?cid=${this.streamKey}`;
    } else {
      this.ingestionUrl = null;
    }

    if (this.verbose) {
      logger.setVerbose(true);
    }
  }

  /**
   * Format timestamp to Google's expected format: YYYY-MM-DDTHH:MM:SS.mmm
   * (no 'Z' suffix, no timezone offset).
   * Accepts a Date object, an ISO string (with or without trailing 'Z'), or undefined (auto-generates current time).
   */
  _formatTimestamp(timestamp) {
    if (timestamp instanceof Date) {
      return timestamp.toISOString().slice(0, -1); // Remove the 'Z'
    }
    if (timestamp && !timestamp.endsWith('Z')) {
      return timestamp;
    }
    const date = timestamp ? new Date(timestamp) : new Date();
    const iso = date.toISOString();
    return iso.slice(0, -1); // Remove the 'Z'
  }

  /**
   * Build the request URL with seq param
   */
  _buildRequestUrl(seq) {
    const parsedUrl = new URL(this.ingestionUrl);
    parsedUrl.searchParams.set('seq', seq.toString());
    return parsedUrl;
  }

  /**
   * Build caption body for a single caption.
   * Format: timestamp (with optional region/cue) on one line, text on next line.
   * Modify this method to change the caption format globally.
   */
  _buildCaptionBody(timestamp, text, useRegion = false) {
    const ts = this._formatTimestamp(timestamp);
    const regionCue = (useRegion && this.region && this.cue) ? ` region:${this.region}#${this.cue}` : '';
    return `${ts}${regionCue}\n${text}`;
  }

  /**
   * Internal method to send POST request
   */
  _sendPost(body, seq) {
    return new Promise((resolve, reject) => {
      let url;
      try {
        url = this._buildRequestUrl(seq);
      } catch (err) {
        reject(new ConfigError(`Invalid ingestion URL: ${this.ingestionUrl}`));
        return;
      }

      const options = {
        hostname: url.hostname,
        port: url.port || 80,
        path: url.pathname + url.search,
        method: 'POST',
        headers: {
          'Content-Type': 'text/plain',
          'Content-Length': Buffer.byteLength(body)
        }
      };

      logger.debug(`POST ${options.path}`);
      logger.debug(`Host: ${options.hostname}`);
      logger.debug(`Content-Type: ${options.headers['Content-Type']}`);
      logger.debug(`Content-Length: ${options.headers['Content-Length']}`);
      logger.debug(`\n${body}`);

      const req = http.request(options, (res) => {
        let data = '';
        res.on('data', (chunk) => { data += chunk; });
        res.on('end', () => {
          const serverTimestamp = data.trim() || null;
          logger.debug(`Response [${res.statusCode}]: ${data || '(empty)'}`);
          resolve({
            statusCode: res.statusCode,
            response: data,
            serverTimestamp
          });
        });
      });

      req.on('error', (err) => {
        logger.error(`Network error: ${err.message}`);
        reject(new NetworkError(`Failed to send: ${err.message}`));
      });

      req.write(body);
      req.end();
    });
  }

  /**
   * Start the caption sender. Must be called before sending captions.
   * @returns {YoutubeLiveCaptionSender} The sender instance for chaining
   * @throws {Error} If already started
   * @example
   * sender.start();
   */
  start() {
    this.isStarted = true;
    logger.info(`Caption sender started`);
    return this;
  }

  /**
   * Send a single caption to the YouTube Live stream.
   * @param {string} text - Caption text (use <br> for line breaks within the caption)
   * @param {string|Date} [timestamp] - Timestamp as a Date object, an ISO string (`YYYY-MM-DDTHH:MM:SS.mmm`),
   *   or an ISO string with trailing 'Z' (auto-converted). Auto-generated if not provided.
   *   Must be within 60 seconds of the server's current time.
   * @returns {Promise<Object>} Result object
   * @returns {number} return.sequence - The sequence number used for this caption
   * @returns {string} return.timestamp - The formatted timestamp sent with the caption
   * @returns {number} return.statusCode - HTTP status code from the server
   * @returns {string} return.response - Raw response body from the server
   * @returns {string|null} return.serverTimestamp - Server timestamp if returned
   * @throws {ValidationError} If sender not started or text is invalid
   * @throws {ConfigError} If no ingestion URL configured
   * @throws {NetworkError} If the request fails
   * @example
   * const result = await sender.send('Hello, world!');
   * console.log(result.sequence); // 0
   * @example
   * // With ISO string timestamp
   * await sender.send('Custom time', '2024-01-15T12:00:00.000');
   * @example
   * // With Date object
   * await sender.send('Custom time', new Date());
   */
  async send(text, timestamp) {
    const result = await this.sendBatch([{ text, timestamp }]);
    return {
      sequence: result.sequence,
      timestamp: this._formatTimestamp(timestamp),
      statusCode: result.statusCode,
      response: result.response,
      serverTimestamp: result.serverTimestamp
    };
  }

  /**
   * Add a caption to the internal queue for later batch sending with sendBatch().
   * @param {string} text - Caption text (use <br> for line breaks within the caption)
   * @param {string|Date} [timestamp] - Timestamp as a Date object, an ISO string (`YYYY-MM-DDTHH:MM:SS.mmm`),
   *   or an ISO string with trailing 'Z' (auto-converted). Auto-generated at send time if not provided.
   *   Must be within 60 seconds of the server's current time.
   * @returns {number} Number of captions currently in queue after adding
   * @throws {ValidationError} If text is empty or not a string
   * @example
   * sender.construct('First caption');
   * sender.construct('Second caption', '2024-01-15T12:00:00.500');
   * sender.construct('Third caption', new Date());
   * await sender.sendBatch(); // Sends all queued captions
   */
  construct(text, timestamp) {
    if (!text || typeof text !== 'string') {
      throw new ValidationError('Caption text is required and must be a string.', 'text');
    }

    this._queue.push({ text, timestamp: timestamp || null });
    logger.debug(`Queued caption: "${text.substring(0, 30)}..." (${this._queue.length} in queue)`);
    return this._queue.length;
  }

  /**
   * Get a copy of the current caption queue.
   * @returns {Array<{text: string, timestamp: string|null}>} Copy of the queue array
   * @example
   * sender.construct('Caption 1');
   * sender.construct('Caption 2');
   * const queue = sender.getQueue();
   * console.log(queue.length); // 2
   */
  getQueue() {
    return [...this._queue];
  }

  /**
   * Clear all captions from the internal queue.
   * @returns {number} Number of captions that were cleared
   * @example
   * sender.construct('Caption 1');
   * sender.construct('Caption 2');
   * const cleared = sender.clearQueue(); // Returns 2
   * console.log(sender.getQueue().length); // 0
   */
  clearQueue() {
    const count = this._queue.length;
    this._queue = [];
    logger.debug(`Cleared ${count} caption(s) from queue`);
    return count;
  }

  /**
   * Send multiple captions in a single POST request.
   * If no captions array is provided, sends the internal queue built with construct() and clears it.
   * @param {Array<{text: string, timestamp?: string|Date}>} [captions] - Array of caption objects. If omitted, uses internal queue.
   *   Each `timestamp` may be a Date object, an ISO string (`YYYY-MM-DDTHH:MM:SS.mmm`), or omitted (auto-generated, 100ms apart).
   * @returns {Promise<Object>} Result object
   * @returns {number} return.sequence - The sequence number used for this batch
   * @returns {number} return.count - Number of captions sent
   * @returns {number} return.statusCode - HTTP status code from the server
   * @returns {string} return.response - Raw response body from the server
   * @returns {string|null} return.serverTimestamp - Server timestamp if returned
   * @throws {ValidationError} If sender not started or no captions to send
   * @throws {ConfigError} If no ingestion URL configured
   * @throws {NetworkError} If the request fails
   * @example
   * // Option 1: Pass array directly (string or Date timestamps)
   * await sender.sendBatch([
   *   { text: 'First caption' },
   *   { text: 'Second caption', timestamp: '2024-01-15T12:00:00.500' },
   *   { text: 'Third caption', timestamp: new Date() }
   * ]);
   * @example
   * // Option 2: Use construct() then sendBatch()
   * sender.construct('First caption');
   * sender.construct('Second caption');
   * await sender.sendBatch(); // Sends queue and clears it
   */
  async sendBatch(captions) {
    if (!this.isStarted) {
      throw new ValidationError('Sender not started. Call start() first.', 'isStarted');
    }

    if (!this.ingestionUrl) {
      throw new ConfigError('No ingestion URL configured.');
    }

    // Use internal queue if no captions provided
    const captionsToSend = captions || this._queue;

    if (!Array.isArray(captionsToSend) || captionsToSend.length === 0) {
      throw new ValidationError('No captions to send. Use construct() to queue captions first, or pass an array.', 'captions');
    }

    const seq = this.sequence;
    const lines = [];

    // Build body with each caption
    for (let i = 0; i < captionsToSend.length; i++) {
      const caption = captionsToSend[i];

      if (!caption.text || typeof caption.text !== 'string') {
        throw new ValidationError(`Caption at index ${i} must have a text string.`, 'captions');
      }

      // Auto-generate timestamp if not provided, spacing them 100ms apart
      let timestamp = caption.timestamp;
      if (!timestamp) {
        const now = new Date();
        now.setMilliseconds(now.getMilliseconds() + (i * 100));
        timestamp = now.toISOString();
      }

      lines.push(this._buildCaptionBody(timestamp, caption.text, this.useRegion));
    }

    const body = lines.join('\n') + '\n';
    logger.debug(`Batch body (${lines.length} captions):\n${body}`);
    const result = await this._sendPost(body, seq);

    // Clear queue if we used it
    if (!captions) {
      this._queue = [];
    }

    if (result.statusCode >= 200 && result.statusCode < 300) {
      this.sequence++;
      logger.success(`Sent batch #${seq}: ${captionsToSend.length} caption(s)`);
    } else {
      logger.warn(`Batch #${seq} sent with status ${result.statusCode}`);
    }

    return {
      sequence: seq,
      count: captionsToSend.length,
      statusCode: result.statusCode,
      response: result.response,
      serverTimestamp: result.serverTimestamp
    };
  }

  /**
   * Send a heartbeat (empty POST) to verify the connection is working.
   * Can also be used for clock synchronization via the returned server timestamp.
   * Note: Heartbeat does NOT increment the sequence number per Google's spec.
   * @returns {Promise<Object>} Result object
   * @returns {number} return.sequence - The current sequence number (not incremented)
   * @returns {number} return.statusCode - HTTP status code from the server
   * @returns {string|null} return.serverTimestamp - Server timestamp for clock sync
   * @throws {ValidationError} If sender not started
   * @throws {ConfigError} If no ingestion URL configured
   * @throws {NetworkError} If the request fails
   * @example
   * const result = await sender.heartbeat();
   * if (result.statusCode === 200) {
   *   console.log('Connection OK, server time:', result.serverTimestamp);
   * }
   */
  async heartbeat() {
    if (!this.isStarted) {
      throw new ValidationError('Sender not started. Call start() first.', 'isStarted');
    }

    if (!this.ingestionUrl) {
      throw new ConfigError('No ingestion URL configured.');
    }

    const seq = this.sequence;
    const body = ''; // Empty body for heartbeat

    const result = await this._sendPost(body, seq);
    // Note: heartbeat does NOT increment sequence per Google's spec
    // (only new caption data increments)

    if (result.statusCode >= 200 && result.statusCode < 300) {
      logger.success(`Heartbeat #${seq} OK`);
    } else {
      logger.warn(`Heartbeat #${seq} returned status ${result.statusCode}`);
    }

    return {
      sequence: seq,
      statusCode: result.statusCode,
      serverTimestamp: result.serverTimestamp
    };
  }

  /**
   * Stop the caption sender and cleanup.
   * @returns {YoutubeLiveCaptionSender} The sender instance for chaining
   * @example
   * sender.end();
   */
  end() {
    this.isStarted = false;
    logger.info(`Caption sender stopped. Total captions sent: ${this.sequence}`);
    return this;
  }

  /**
   * Get the current sequence number.
   * @returns {number} Current sequence number
   * @example
   * const seq = sender.getSequence();
   */
  getSequence() {
    return this.sequence;
  }

  /**
   * Set the sequence number manually.
   * @param {number} seq - New sequence number
   * @returns {YoutubeLiveCaptionSender} The sender instance for chaining
   * @example
   * sender.setSequence(100);
   */
  setSequence(seq) {
    this.sequence = seq;
    return this;
  }

  /**
   * Test function that sends the exact example body from Google's docs.
   */
  async sendTest() {
    if (!this.isStarted) {
      throw new ValidationError('Sender not started. Call start() first.', 'isStarted');
    }
    if (!this.ingestionUrl) {
      throw new ConfigError('No ingestion URL configured.');
    }

    // Generate current timestamps (must be within 60 seconds of server time)
    const now = new Date();
    const ts1 = this._formatTimestamp(now.toISOString());
    now.setMilliseconds(now.getMilliseconds() + 100);
    const ts2 = this._formatTimestamp(now.toISOString());

    const body = `${ts1} region:reg1#cue1
HELLO
${ts2} region:reg1#cue1
WORLD
`;

    const seq = this.sequence;
    const result = await this._sendPost(body, seq);

    if (result.statusCode >= 200 && result.statusCode < 300) {
      this.sequence++;
      logger.success(`Test sent #${seq}`);
    } else {
      logger.warn(`Test #${seq} sent with status ${result.statusCode}`);
    }

    return result;
  }
}

export { YoutubeLiveCaptionSender, DEFAULT_YOUTUBE_URL };

File: python-packages/lcyt/lcyt/sender.py
"""YouTube Live Caption Sender."""

import http.client
import logging
import time
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any
from urllib.parse import urlparse

from .config import DEFAULT_BASE_URL, build_ingestion_url, LCYTConfig
from .errors import NetworkError, ValidationError

logger = logging.getLogger("lcyt")


@dataclass
class Caption:
    """A single caption with text and optional timestamp."""

    text: str
    timestamp: str | datetime | int | float | None = None


@dataclass
class SendResult:
    """Result of a send operation."""

    sequence: int
    status_code: int
    response: str
    server_timestamp: str | None = None
    timestamp: str | None = None
    count: int | None = None


class YoutubeLiveCaptionSender:
    """Send live captions to YouTube streams.

    This class provides methods to send captions to YouTube's live caption
    ingestion endpoint. It supports single captions, batch sending, and
    queue-based workflows.

    Example:
        >>> sender = YoutubeLiveCaptionSender(stream_key="YOUR_KEY")
        >>> sender.start()
        >>> result = sender.send("Hello, world!")
        >>> sender.end()
    """

    def __init__(
        self,
        stream_key: str | None = None,
        base_url: str = DEFAULT_BASE_URL,
        ingestion_url: str | None = None,
        region: str = "reg1",
        cue: str = "cue1",
        use_region: bool = False,
        sequence: int = 0,
        use_sync_offset: bool = False,
        verbose: bool = False,
    ):
        """Initialize the caption sender.

        Args:
            stream_key: YouTube stream key (cid parameter).
            base_url: Base URL for the ingestion endpoint.
            ingestion_url: Full ingestion URL (overrides stream_key + base_url).
            region: Region identifier for captions.
            cue: Cue identifier for captions.
            use_region: Whether to include region/cue in caption body.
            sequence: Starting sequence number.
            use_sync_offset: Apply syncOffset to auto-generated timestamps.
                             Set automatically to True after calling sync().
            verbose: Enable debug logging.
        """
        self._stream_key = stream_key
        self._base_url = base_url
        self._ingestion_url = ingestion_url
        self._region = region
        self._cue = cue
        self._use_region = use_region
        self._sequence = sequence
        self._sync_offset: int = 0   # Clock offset in ms (positive = server ahead)
        self._use_sync_offset = use_sync_offset
        self._verbose = verbose

        self._queue: list[Caption] = []
        self._started = False

        if verbose:
            logging.basicConfig(level=logging.DEBUG)
            logger.setLevel(logging.DEBUG)

    # ------------------------------------------------------------------
    # Internal time helper
    # ------------------------------------------------------------------

    def _now_ms(self) -> float:
        """Return current time in epoch milliseconds, adjusted by sync offset."""
        return time.time() * 1000 + (self._sync_offset if self._use_sync_offset else 0)

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self) -> "YoutubeLiveCaptionSender":
        """Initialize the sender. Must be called before sending captions.

        Returns:
            Self for method chaining.

        Raises:
            ValidationError: If stream_key or ingestion_url is not set.
        """
        if self._ingestion_url:
            self._url = self._ingestion_url
        elif self._stream_key:
            config = LCYTConfig(
                stream_key=self._stream_key,
                base_url=self._base_url,
            )
            self._url = build_ingestion_url(config)
        else:
            raise ValidationError(
                "Either stream_key or ingestion_url must be provided",
                field="stream_key",
            )

        self._started = True
        logger.debug(f"Sender started with URL: {self._url}")
        return self

    def end(self) -> "YoutubeLiveCaptionSender":
        """Stop the sender and cleanup.

        Returns:
            Self for method chaining.
        """
        self._started = False
        self._queue.clear()
        logger.debug(f"Caption sender stopped. Total captions sent: {self._sequence}")
        return self

    # ------------------------------------------------------------------
    # Sending
    # ------------------------------------------------------------------

    def send(self, text: str, timestamp: str | datetime | int | float | None = None) -> SendResult:
        """Send a single caption.

        Args:
            text: Caption text to send.
            timestamp: Optional timestamp. Accepted forms:

                - ``datetime`` object (timezone-aware or naive UTC)
                - ``int``/``float`` >= 1000: Unix epoch in seconds (``time.time()`` style)
                - ``int``/``float`` < 1000 or negative: relative offset in seconds from now
                  (e.g. ``-2`` = 2 seconds ago). Sync offset is applied when enabled.
                - ISO string ``YYYY-MM-DDTHH:MM:SS.mmm`` (used as-is)
                - ISO string with trailing ``Z`` or ``+00:00`` (auto-stripped)
                - ``None``: auto-generated current time (sync offset applied if enabled)

        Returns:
            SendResult. The ``timestamp`` field is set to the formatted string sent.

        Raises:
            ValidationError: If sender not started or text is empty.
            NetworkError: If HTTP request fails.
        """
        self._ensure_started()
        if not text:
            raise ValidationError("Caption text cannot be empty", field="text")

        result = self.send_batch([Caption(text=text, timestamp=timestamp)])
        # Populate the timestamp field to mirror Node.js send() return shape
        if timestamp is not None:
            result.timestamp = self._format_timestamp(timestamp)
        return result

    def construct(self, text: str, timestamp: str | datetime | int | float | None = None) -> int:
        """Queue a caption for batch sending.

        Args:
            text: Caption text to queue.
            timestamp: Optional timestamp. Accepts the same forms as ``send()``.

        Returns:
            Current queue length.

        Raises:
            ValidationError: If sender not started or text is empty/not a string.
        """
        self._ensure_started()
        if not text or not isinstance(text, str):
            raise ValidationError("Caption text is required and must be a string.", field="text")
        self._queue.append(Caption(text=text, timestamp=timestamp))
        logger.debug(f"Caption queued, queue length: {len(self._queue)}")
        return len(self._queue)

    def send_batch(self, captions: list[Caption] | None = None) -> SendResult:
        """Send a batch of captions.

        Args:
            captions: List of captions to send. If None, sends queued captions.

        Returns:
            SendResult with sequence, count, status code, and response.

        Raises:
            ValidationError: If sender not started or no captions to send.
            NetworkError: If HTTP request fails.
        """
        self._ensure_started()

        if captions is None:
            captions = self._queue.copy()
            self._queue.clear()

        if not captions:
            raise ValidationError("No captions to send")

        body_parts: list[str] = []
        base_time_ms = self._now_ms()

        for i, caption in enumerate(captions):
            if caption.timestamp:
                ts = self._format_timestamp(caption.timestamp)
            else:
                # Space captions 100ms apart if no timestamp provided.
                caption_time = datetime.fromtimestamp(
                    (base_time_ms + i * 100) / 1000, tz=timezone.utc
                )
                ts = self._format_timestamp(caption_time)

            body_parts.append(self._build_caption_body(ts, caption.text))

        body = "\n".join(body_parts) + "\n"
        sent_sequence = self._sequence
        result = self._send_post(body, sent_sequence)

        if 200 <= result.status_code < 300:
            self._sequence += 1
            logger.debug(f"Sent batch #{sent_sequence}: {len(captions)} caption(s)")
        else:
            logger.debug(f"Batch #{sent_sequence} returned status {result.status_code}")

        return SendResult(
            sequence=sent_sequence,
            count=len(captions),
            status_code=result.status_code,
            response=result.response,
            server_timestamp=result.server_timestamp,
        )

    def heartbeat(self) -> SendResult:
        """Send a heartbeat (empty POST) to verify connection.

        The heartbeat does not increment the sequence number per Google's spec.

        Returns:
            SendResult with status code and server timestamp.

        Raises:
            ValidationError: If sender not started.
            NetworkError: If HTTP request fails.
        """
        self._ensure_started()
        result = self._send_post("", self._sequence)
        if 200 <= result.status_code < 300:
            logger.debug(f"Heartbeat #{self._sequence} OK")
        else:
            logger.debug(f"Heartbeat #{self._sequence} returned status {result.status_code}")
        return SendResult(
            sequence=self._sequence,
            status_code=result.status_code,
            response=result.response,
            server_timestamp=result.server_timestamp,
        )

    def sync(self) -> dict:
        """Synchronize the local clock with YouTube's server clock (NTP-style).

        Sends a heartbeat, measures round-trip time, computes the midpoint
        estimate of server time, and stores the offset as ``sync_offset``.
        Automatically enables ``use_sync_offset`` so future auto-generated
        timestamps are corrected.

        Returns:
            dict with keys:
                ``sync_offset`` (int ms), ``round_trip_time`` (int ms),
                ``server_timestamp`` (str | None), ``status_code`` (int).

        Raises:
            ValidationError: If sender not started.
            NetworkError: If HTTP request fails.
        """
        self._ensure_started()

        t1_wall_ms = time.time() * 1000
        result = self.heartbeat()
        t2_wall_ms = time.time() * 1000
        rtt_ms = int(t2_wall_ms - t1_wall_ms)

        if not result.server_timestamp:
            logger.debug("No server timestamp in heartbeat response — syncOffset not updated")
            return {
                "sync_offset": self._sync_offset,
                "round_trip_time": rtt_ms,
                "server_timestamp": None,
                "status_code": result.status_code,
            }

        # Parse server timestamp (format: YYYY-MM-DDTHH:MM:SS.mmm — no Z, treat as UTC)
        server_time_ms = (
            datetime.fromisoformat(result.server_timestamp)
            .replace(tzinfo=timezone.utc)
            .timestamp()
        ) * 1000

        local_estimate_ms = (t1_wall_ms + t2_wall_ms) / 2
        self._sync_offset = round(server_time_ms - local_estimate_ms)
        self._use_sync_offset = True

        logger.debug(f"Synced: offset {self._sync_offset}ms, RTT {rtt_ms}ms")

        return {
            "sync_offset": self._sync_offset,
            "round_trip_time": rtt_ms,
            "server_timestamp": result.server_timestamp,
            "status_code": result.status_code,
        }

    def send_test(self) -> SendResult:
        """Send a test payload using current timestamps.

        Uses the ``region:reg1#cue1`` format from Google's documentation.

        Returns:
            SendResult with status code and response.

        Raises:
            ValidationError: If sender not started.
            NetworkError: If HTTP request fails.
        """
        self._ensure_started()

        now_ms = self._now_ms()
        ts1 = self._format_timestamp(datetime.fromtimestamp(now_ms / 1000, tz=timezone.utc))
        ts2 = self._format_timestamp(
            datetime.fromtimestamp((now_ms + 100) / 1000, tz=timezone.utc)
        )

        body = (
            f"{ts1} region:reg1#cue1\n"
            "HELLO\n"
            f"{ts2} region:reg1#cue1\n"
            "WORLD\n"
        )

        sent_sequence = self._sequence
        result = self._send_post(body, sent_sequence)

        if 200 <= result.status_code < 300:
            self._sequence += 1
            logger.debug(f"Test sent #{sent_sequence}")
        else:
            logger.debug(f"Test #{sent_sequence} returned status {result.status_code}")

        return result

    # ------------------------------------------------------------------
    # Queue helpers
    # ------------------------------------------------------------------

    def get_queue(self) -> list[Caption]:
        """Get a copy of the current caption queue."""
        return self._queue.copy()

    def clear_queue(self) -> int:
        """Clear all queued captions.

        Returns:
            Number of captions cleared.
        """
        count = len(self._queue)
        self._queue.clear()
        logger.debug(f"Cleared {count} caption(s) from queue")
        return count

    # ------------------------------------------------------------------
    # Sequence management
    # ------------------------------------------------------------------

    def get_sequence(self) -> int:
        """Get the current sequence number."""
        return self._sequence

    def set_sequence(self, sequence: int) -> "YoutubeLiveCaptionSender":
        """Set the sequence number manually."""
        self._sequence = sequence
        return self

    # ------------------------------------------------------------------
    # Sync offset management
    # ------------------------------------------------------------------

    def get_sync_offset(self) -> int:
        """Get the current sync offset in milliseconds.

        Returns:
            Clock offset in ms (positive = server ahead of local).
        """
        return self._sync_offset

    def set_sync_offset(self, offset: int) -> "YoutubeLiveCaptionSender":
        """Set the sync offset manually (e.g. to restore a previously computed offset).

        Args:
            offset: Clock offset in ms.

        Returns:
            Self for method chaining.
        """
        self._sync_offset = offset
        return self

    @property
    def is_started(self) -> bool:
        """Check if sender is started."""
        return self._started

    # ------------------------------------------------------------------
    # Private implementation
    # ------------------------------------------------------------------

    def _ensure_started(self) -> None:
        if not self._started:
            raise ValidationError("Sender not started. Call start() first.")

    def _format_timestamp(self, timestamp: str | datetime | int | float) -> str:
        """Format timestamp for YouTube API.

        YouTube expects: YYYY-MM-DDTHH:MM:SS.mmm (milliseconds, no timezone suffix)

        Accepted inputs:
        - ``datetime``: converted via isoformat()
        - ``int``/``float`` >= 1000: Unix epoch in **seconds** (time.time() convention)
        - ``int``/``float`` < 1000 or negative: relative seconds offset from now
          (sync offset applied when use_sync_offset is True)
        - ISO string with or without trailing 'Z' or '+00:00'
        """
        if isinstance(timestamp, datetime):
            timestamp = timestamp.isoformat()
        elif isinstance(timestamp, (int, float)):
            if timestamp < 1000:
                # Relative offset in seconds from now (sync offset applied)
                dt = datetime.fromtimestamp(
                    (self._now_ms() + timestamp * 1000) / 1000, tz=timezone.utc
                )
            else:
                # Unix epoch in seconds (time.time() convention)
                dt = datetime.fromtimestamp(timestamp, tz=timezone.utc)
            timestamp = dt.isoformat()

        # Strip trailing Z
        if timestamp.endswith("Z"):
            timestamp = timestamp[:-1]
        # Strip +00:00 timezone offset
        if "+" in timestamp:
            timestamp = timestamp.split("+")[0]
        # Truncate to milliseconds (3 decimal places) — YouTube rejects microseconds
        if "." in timestamp:
            base, frac = timestamp.rsplit(".", 1)
            timestamp = f"{base}.{frac[:3]}"
        return timestamp

    def _build_caption_body(self, timestamp: str, text: str) -> str:
        """Build the caption body for a single caption.

        Format (with region): ``{ts} region:{region}#{cue}\\n{text}``
        Format (without):     ``{ts}\\n{text}``
        """
        if self._use_region:
            return f"{timestamp} region:{self._region}#{self._cue}\n{text}"
        return f"{timestamp}\n{text}"

    def _build_request_url(self, sequence: int) -> str:
        separator = "&" if "?" in self._url else "?"
        return f"{self._url}{separator}seq={sequence}"

    def _send_post(self, body: str, sequence: int) -> SendResult:
        """Send HTTP POST request to YouTube."""
        url = self._build_request_url(sequence)
        parsed = urlparse(url)

        logger.debug(f"POST {url}")
        logger.debug(f"Body: {body!r}")

        try:
            if parsed.scheme == "https":
                conn = http.client.HTTPSConnection(parsed.netloc, timeout=30)
            else:
                conn = http.client.HTTPConnection(parsed.netloc, timeout=30)

            path = parsed.path
            if parsed.query:
                path = f"{path}?{parsed.query}"

            encoded = body.encode("utf-8")
            headers = {
                "Content-Type": "text/plain",
                "Content-Length": str(len(encoded)),
            }

            conn.request("POST", path, encoded, headers)
            response = conn.getresponse()
            response_body = response.read().decode("utf-8")
            conn.close()

            logger.debug(f"Response: {response.status} {response_body}")

            server_timestamp = response_body.strip() if response_body.strip() else None

            return SendResult(
                sequence=sequence,
                status_code=response.status,
                response=response_body,
                server_timestamp=server_timestamp,
            )

        except Exception as e:
            raise NetworkError(f"HTTP request failed: {e}") from e
